src/Elements/types.jl

```diff
@@ -0,0 +1,180 @@
+                    :typealias, :global,   :symbol, :tuple, :aside]
+
+
+## Config Nodes
+abstract ConfigNode
+
+type Config <: ConfigNode
+    objfilter               :: Any
+    style_obj               :: ASCIIString
+
+    const defaults = Dict{Symbol, Any}(
+        :objfilter                  => CATEGORIES,
+        :style_obj                  => "####",
+        )
+
+    Config(; args...) = update_config!(new(), merge(defaults, Dict(args)))
+end
+config(; cargs...) = PreConfig(; cargs...)
+
+type PreConfig <: ConfigNode
+    cargs :: Dict{Symbol, Any}
+
+    PreConfig(; cargs...) = new(Dict(cargs))
+end
+
+function update_config!(config::Config, args::Dict)
+    const fields = fieldnames(Config)
+    for (k, v) in args
+        k in fields                                                ?
+            setfield!(config, k, convert(fieldtype(Config, k), v)) :
+            throw(ArgumentError("'Config'. Invalid setting: '$(k) = $(v)'."))
+    end
+
+    for k in [:style_obj]
+        getfield(config, k) in HEADERS ||
+                error("""Invalid style : config-item `$k -> $(getfield(config, k))`.
+                      Valid values: [$(join(MDSTYLETAGS, ", "))].""")
+    end
+    return config
+end
+update_config!(config::Config, args...) = update_config!(config, Dict(args))
+
+
+## Main structure nodes
+abstract SectionNode
+abstract PageNode
+abstract ContentNode
+
+immutable NullSection <: SectionNode end
+immutable NullPage    <: PageNode    end
+
+# Document
+type Document
+    name        :: AbstractString
+    config      :: ConfigNode
+    meta        :: Dict{Symbol, Any}
+    children    :: Vector{SectionNode}
+end
+document(name::AbstractString, conf::PreConfig, children...) =
+        document(name, conf, Dict(), [children...])
+document(name::AbstractString, children...) = document(name, PreConfig(), children...)
+
+function document(name::AbstractString, conf::PreConfig, meta::Dict, children::Vector)
+    this = isempty(conf.cargs)                       ?
+            Document(name, Config(), meta, children) :
+            Document(name, update_config!(Config(), deepcopy(conf.cargs)), meta, children)
+    isempty(children) || postprocess!(this, children)
+    return this
+end
+
+# Section
+type Section <: SectionNode
+    name            :: AbstractString
+    config          :: ConfigNode
+    meta            :: Dict{Symbol, Any}
+    parent          :: Union(Document, SectionNode)
+    children        :: Vector{Union(SectionNode, PageNode)}
+end
+section(name::AbstractString, conf::PreConfig, children...) =
+        Section(name, conf, Dict(:autogenerate => (false, nothing)), NullSection(), [children...])
+section(name::AbstractString, children...) = section(name, PreConfig(), children...)
+
+# Page
+type Page <: PageNode
+    name            :: AbstractString
+    config          :: ConfigNode
+    meta            :: Dict{Symbol, Any}
+    parent          :: SectionNode
+    children        :: Vector{ContentNode}
+end
```

Comment:
I was thinking about that but a couple of days ago you where quite strict against something similar - just can't find it yet ;)

Anyway, I can give that a try would probably make sense.

> Should name just be a field in the meta Dict?

Not sure: as I said in my first note this names are essential - but probably it could be if it does not get overwritten.

**IMPORTANT:** I was looking for an immutable dict a couple of days ago and found this: [FunctionalCollections.jl](https://github.com/JuliaLang/FunctionalCollections.jl) but I'm not sure if you want to make an extra dependency. I was also thinking to extract the only the code for the immutable dictionar into one file and maybe we could include such?  [PersistentHashMap](https://github.com/JuliaLang/FunctionalCollections.jl/blob/master/src/PersistentHashMap.jl)  I had started it once so I think it should be possible?